sstables/sstable_set.cc

```diff
@@ -768,6 +768,393 @@ sstable_set_impl::create_single_key_sstable_reader(
+        }
+    };
+
+    const peeked_reader_cmp _peeked_cmp;
+
+    // operator() returns a mutation_fragment_batch, which is a range (a pair of iterators);
+    // this is where the actual data is stored, i.e. the range points to _current_batch.
+    merger_vector<mutation_fragment> _current_batch;
+
+    // _unpeeked_readers stores readers for which we don't know the next fragment that they'll return.
+    // Before we return the next batch of fragments, we must peek all readers here (and move them to
+    // the _peeked_readers heap), since they might contain fragments with smaller positions than the
+    // currently peeked readers.
+    merger_vector<reader_iterator> _unpeeked_readers;
+
+    // In forwarding mode, after a reader returns end-of-stream, we save it in _halted_readers,
+    // a min-heap sorted by the max_position()s of the readers' sstables.
+    // After each returned batch of fragments we compare the position of this batch to the
+    // max_position() of the heap's front; if the max_position() is smaller, it means that
+    // this reader won't return any more fragments, so we can remove it.
+    // We restore all halted readers that weren't removed when we fast-forward, since they
+    // might start returning fragments again.
+    // _halted_readers doesn't serve any purpose when not in forwarding mode, because then
+    // readers always return end-of-partition before end-of-stream, which is a signal that
+    // we can remove the reader immediately.
+    merger_vector<reader_iterator> _halted_readers;
+
+    // Used to compare readers stored in _halted_readers.
+    struct halter_reader_cmp {
+        const position_in_partition::less_compare _less;
+
+        explicit halter_reader_cmp(position_in_partition::less_compare cmp)
+            : _less(std::move(cmp)) {}
+
+        bool operator()(reader_iterator a, reader_iterator b) {
+            // Boost heaps are max-heaps, but we want a min-heap, so invert the comparison.
+            return _less(b->sst->max_position(), a->sst->max_position());
+        }
+    };
+
+    const halter_reader_cmp _halted_reader_cmp;
+
+    // In forwarding mode, this is the right-end of the position range being currently queried;
+    // initially it's set to `before_all_clustered_rows` and updated on `fast_forward_to`.
+    // We use it when popping readers from _reader_queue so that we don't prematurely pop
+    // readers that only contain fragments from greater ranges.
+    // In non-forwarding mode _pr_end is always equal to `after_all_clustered_rows`.
+    position_in_partition _pr_end;
+
+    // In forwarding mode, _forwarded_to remembers the last range we were forwarded to.
+    // We need this because we're opening new readers in the middle of the partition query:
+    // after the new reader returns its initial partition-start, we immediately forward it
+    // to this range.
+    std::optional<position_range> _forwarded_to;
```

Comment:
`_position_range` would be a better name.

Reply:
Why? `_position_range` doesn't tell the reader of code anything except what they already know from the type.
`_forwarded_to` reflects the semantics of the variable. It is the range that we have been forwarded to.